### PR TITLE
reindeer: init at unstable-2023-05-18

### LINKED ISSUE #232693: Package request: reindeer
**Project description**

> "Reindeer is a tool which takes Rust Cargo dependencies and generates Buck build rules."
>
> -- https://facebookincubator/reindeer

Related issue: https://github.com/NixOS/nixpkgs/issues/226677
Related PR: https://github.com/NixOS/nixpkgs/pull/232471

**Metadata**

* homepage URL: https://github.com/facebookincubator/reindeer
* source URL: https://github.com/facebookincubator/reindeer
* license: mit
* platforms: linux and darwin (might also work on other platforms)

### PR DESCRIPTION
**Pull request by nickgerace**
###### Description of changes

This PR adds [reindeer](https://github.com/facebookincubator/reindeer) with the ability to auto update using `nix-update-script`.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Additional notes

- Fixes #232693
- Did not add anything to the release notes as `buck2` is not a service

### CODE REVIEW

File: pkgs/development/tools/reindeer/default.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  meta = with lib;
+    {
+      description = "Reindeer is a tool which takes Rust Cargo dependencies and generates Buck build rules";
+      homepage = "https://github.com/facebookincubator/reindeer";
+      license = with licenses; [ mit ];
+      maintainers = with maintainers; [ nickgerace ];
+    };
+
+  passthru.updateScript = nix-update-script { };
```
Comment (nickgerace) [inline pkgs/development/tools/reindeer/default.nix:49]: @figsoda thanks for yourself on the other PR! I navigated to `update.nix` and other examples to try to find how to use this since [there are not docs for this yet](https://github.com/NixOS/nixpkgs/issues/216321). I _believe_ I might not need to do more here, but I want to be sure. I'd like the package to auto update the following attributes...

- `version` (update to `unstable-YYYY-MM-DD`)
- `src.rev` (update to latest commit sha on `main` for the source repo)
- `src.sha256` (update accordingly, based on `src.rev`)

...and then `Cargo.lock` needs to be regenerated. I did [something similar in my `buck2` branch here](https://github.com/nickgerace/nixpkgs/blob/48feccd207c4c58a91ebdeafbfff402210c9f092/pkgs/development/tools/build-managers/buck2/update.sh#L23-L31), but that uses the `update.sh` pattern.

Does this require any further configuration? My guess is "yes", but I'm still looking.

File: pkgs/development/tools/reindeer/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+, fetchFromGitHub
+, pkg-config
+, openssl
+, libiconv
+, darwin
+, stdenv
+, nix-update-script
+}:
+
+rustPlatform.buildRustPackage rec {
```
Comment (nickgerace) [inline pkgs/development/tools/reindeer/default.nix:12]: Unlike [buck2](https://github.com/NixOS/nixpkgs/pull/232471#issuecomment-15534815400), we can and should use stable Rust.

File: pkgs/development/tools/reindeer/default.nix (hunk 1 above)
Comment (figsoda) [inline pkgs/development/tools/reindeer/default.nix:49]:
```suggestion
  passthru.updateScript = nix-update-script { };

  meta = with lib; {
    description = "Reindeer is a tool which takes Rust Cargo dependencies and generates Buck build rules";
    homepage = "https://github.com/facebookincubator/reindeer";
    license = with licenses; [ mit ];
    maintainers = with maintainers; [ nickgerace ];
  };
```

File: pkgs/development/tools/reindeer/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  postPatch = ''
+    ln -s ${./Cargo.lock} Cargo.lock
+  '';
+
+  doCheck = false;
```
Comment (figsoda) [inline pkgs/development/tools/reindeer/default.nix:31]: please add a comment explaining why

File: pkgs/development/tools/reindeer/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{ lib
+, rustPlatform
+, fetchFromGitHub
+, pkg-config
+, openssl
+, libiconv
+, darwin
+, stdenv
```
Comment (figsoda) [inline pkgs/development/tools/reindeer/default.nix:8]: > [Arguments should be listed in the order they are used, with the exception of `lib`, which always goes first.](https://nixos.org/manual/nixpkgs/unstable/#sec-syntax)

```suggestion
, stdenv
, libiconv
, darwin
```

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -11275,6 +11275,11 @@
     github = "NickCao";
     githubId = 15247171;
   };
+  nickgerace = {
```
Comment (figsoda) [inline maintainers/maintainer-list.nix:11278]: could you put this in a separate commit?

```
maintainers: add nickgerace
reindeer: init at unstable-2023-05-18
```

**Review by figsoda — COMMENTED** [review]
mostly lgtm, just some nitpicks

File: pkgs/development/tools/reindeer/default.nix (hunk 1 above)
Comment (figsoda) [reply pkgs/development/tools/reindeer/default.nix:49]: for some background since there's no docs for it yet, nix-update-script is defined [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/common-updater/nix-update.nix) and uses [nix-update](https://github.com/Mic92/nix-update) under the hood. Just `nix-update-script { }` (roughly equivalent to running `nix-update reindeer`) should be sufficient here without any further configuration

**Comment by nickgerace** [comment]
Addressing your comments! Thank you.

File: pkgs/development/tools/reindeer/default.nix (hunk 3 above)
Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix:31]: This was carryover from the buck2 PR (whose tests were not always passing), so I just tested locally and I think we can remove this!

File: pkgs/development/tools/reindeer/default.nix (hunk 1 above)
Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix]: Huh that is super cool. I can't believe I don't have to do anything else here. Will have to dive deep and learn more about how it works.

These might be naive, but I have a few more questions:

- Will the @r-ryantm bot pick this up for automatic updates? Automatic updates would be ideal.
- Should running `nix-update reindeer` work locally despite the package not being added to nixpkgs yet?

**Comment by nickgerace** [comment]
Other than our thread on `nix-update-script`, all comments have been addressed.

File: pkgs/development/tools/reindeer/default.nix (hunk 1 above)
Comment (figsoda) [reply pkgs/development/tools/reindeer/default.nix]: > Will the @r-ryantm bot pick this up for automatic updates?

Yes

> Should running `nix-update reindeer` work locally despite the package not being added to nixpkgs yet?

Yes, you should also be able to do `nix-shell maintainers/scripts/update.nix --argstr package reindeer`, this is how `passthru.updateScript` is normally used. Running `nix-shell maintainers/scripts/update.nix` will show you more instructions

File: pkgs/development/tools/reindeer/default.nix [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = with lib;
+    {
+      description = "Reindeer is a tool which takes Rust Cargo dependencies and generates Buck build rules";
+      homepage = "https://github.com/facebookincubator/reindeer";
+      license = with licenses; [ mit ];
+      maintainers = with maintainers; [ nickgerace ];
+    };
```
Comment (figsoda) [inline pkgs/development/tools/reindeer/default.nix:47]:
```suggestion
  meta = with lib; {
    description = "Reindeer is a tool which takes Rust Cargo dependencies and generates Buck build rules";
    homepage = "https://github.com/facebookincubator/reindeer";
    license = with licenses; [ mit ];
    maintainers = with maintainers; [ nickgerace ];
  };
```

Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix]: Ah looks like my editor added it back. Going to double check now.

File: pkgs/development/tools/reindeer/default.nix (hunk 1 above)
Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix:39]: I see that makes sense! Looks like I need to provide it the version so that it knows what to do though...
```
nix_update.errors.VersionError: Please specify the version. We can only get the latest version from codeberg/crates.io/gitea/github/gitlab/notabug/pypi/savannah/sourcehut/rubygems projects right now
```

Comment (figsoda) [reply pkgs/development/tools/reindeer/default.nix:39]: this should work then

```nix
  passthru.updateScript = nix-update-script {
    extraArgs = [ "--version=branch" ];
  };
```

Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix:39]: Ah I see! Okay getting closer. I set the version to yesterday and tried to see what would happen when I ran the following command: `nix-shell maintainers/scripts/update.nix --argstr package reindeer`. Seems like I need to tell it that I need the `Cargo.lock` file to be generated because it does not exist.

```
Running update for:
 - reindeer-unstable-2023-05-17: UPDATING ...
 - reindeer-unstable-2023-05-17: ERROR

--- SHOWING ERROR LOG FOR reindeer-unstable-2023-05-17 ----------------------

this derivation will be built:
  /nix/store/b3hm4sfh9pfn6gbxxsq1qyk7jiwnb164-reindeer-unstable-2023-05-18.drv
building '/nix/store/b3hm4sfh9pfn6gbxxsq1qyk7jiwnb164-reindeer-unstable-2023-05-18.drv'...
error: builder for '/nix/store/b3hm4sfh9pfn6gbxxsq1qyk7jiwnb164-reindeer-unstable-2023-05-18.drv' failed with exit code
 1;
       last 5 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/qhiyv03qifwzdhg3z0yhvmzbklz7abfx-source
       > source root is source
       > cp: cannot stat 'source/Cargo.lock': No such file or directory
       > /nix/store/crv0aygayvds8xwyqln2p1zrdsmfw54s-stdenv-darwin/setup: line 145: pop_var_context: head of shell_vari
ables not a function context
```

The [nix-update](https://github.com/Mic92/nix-update) repository has some interesting notes, but nothing related to specific "generate a `Cargo.lock` if it does not exist" behavior.

I also see in the [Rust documentation for nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/rust.section.md) that `allowBuiltinFetchGit` is "for usage outside nixpkgs". I presume that won't help us solve the issue here and that it is unrelated.

Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix:39]: This seems to be the function that it hits... investigating.
https://github.com/Mic92/nix-update/blob/e242e61cec80802334b6d57530196d79f68bb6f1/nix_update/update.py#L159-L178

Comment (figsoda) [reply pkgs/development/tools/reindeer/default.nix:39]: ah I just realized, nix-update only supports it when the repository contains a `Cargo.lock` file. Sorry for the confusion, I think you would have to have a separate update script for now. Or just not have one and manually update it, that would work too

Comment (nickgerace) [reply pkgs/development/tools/reindeer/default.nix:39]: Ah all good! I've filed an issue: https://github.com/Mic92/nix-update/issues/164. Might even take a crack at it, if I find the time.

Alright I'll add an update script and get back to you. Thanks for the help!

Comment (figsoda) [reply pkgs/development/tools/reindeer/default.nix:39]: Thanks!

File: pkgs/development/tools/reindeer/update.sh
Comment (nickgerace) [inline pkgs/development/tools/reindeer/update.sh:1]: Alright, here it is @figsoda! Thanks for the help today and the back and forth.

As an aside, I can see why `nix-update-script` is desired!

**Review by figsoda — APPROVED** [review]

**Comment by nickgerace** [comment]
Since your approval (thank you by the way), I added the `--depth=1` flag to the `git clone` in the update script. Everything else is the same.

**Comment by nickgerace** [comment]
https://github.com/ofborg build reindeer

**Comment by nickgerace** [comment]
@ofborg build reindeer

**Review by figsoda — APPROVED** [review]
